<a href="https://colab.research.google.com/github/ash12hub/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/DS_Unit_1_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 3

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
# TODO
import pandas as pd;
import numpy as np;

In [65]:
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data", header=None);
df.columns = ["A1", "A2", "A3", "A4", "A5", "A6", "A7", "A8", "A9", "A10", "A11", "A12", "A13", "A14", "A15", "A16"]
df.tail(20)

A1     A2      A3 A4 A5  A6  A7     A8 A9 A10  A11 A12 A13    A14  A15 A16
670  b  47.17   5.835  u  g   w   v  5.500  f   f    0   f   g  00465  150   -
671  b  25.83  12.835  u  g  cc   v  0.500  f   f    0   f   g  00000    2   -
672  a  50.25   0.835  u  g  aa   v  0.500  f   f    0   t   g  00240  117   -
673  ?  29.50   2.000  y  p   e   h  2.000  f   f    0   f   g  00256   17   -
674  a  37.33   2.500  u  g   i   h  0.210  f   f    0   f   g  00260  246   -
675  a  41.58   1.040  u  g  aa   v  0.665  f   f    0   f   g  00240  237   -
676  a  30.58  10.665  u  g   q   h  0.085  f   t   12   t   g  00129    3   -
677  b  19.42   7.250  u  g   m   v  0.040  f   t    1   f   g  00100    1   -
678  a  17.92  10.210  u  g  ff  ff  0.000  f   f    0   f   g  00000   50   -
679  a  20.08   1.250  u  g   c   v  0.000  f   f    0   f   g  00000    0   -
680  b  19.50   0.290  u  g   k   v  0.290  f   f    0   f   g  00280  364   -
681  b  27.83   1.000  y  p   d   h  3.000  f   f    0   f   g  00176  537   -
682  b  17.08   3.290  u  g   i   v  0.335  f   f    0   t   g  00140    2   -
683  b  36.42   0.750  y  p   d   v  0.585  f   f    0   f   g  00240    3   -
684  b  40.58   3.290  u  g   m   v  3.500  f   f    0   t   s  00400    0   -
685  b  21.08  10.085  y  p   e   h  1.250  f   f    0   f   g  00260    0   -
686  a  22.67   0.750  u  g   c   v  2.000  f   t    2   t   g  00200  394   -
687  a  25.25  13.500  y  p  ff  ff  2.000  f   t    1   t   g  00200    1   -
688  b  17.92   0.205  u  g  aa   v  0.040  f   f    0   f   g  00280  750   -
689  b  35.00   3.375  u  g   c   h  8.290  f   f    0   t   g  00000    0   -

In [66]:
df = df.replace(["?"], [np.nan])
newdf = df.dropna()
print(newdf.shape)
newdf.head(20)

(653, 16)


A1     A2      A3 A4 A5  A6 A7     A8 A9 A10  A11 A12 A13    A14    A15 A16
0   b  30.83   0.000  u  g   w  v  1.250  t   t    1   f   g  00202      0   +
1   a  58.67   4.460  u  g   q  h  3.040  t   t    6   f   g  00043    560   +
2   a  24.50   0.500  u  g   q  h  1.500  t   f    0   f   g  00280    824   +
3   b  27.83   1.540  u  g   w  v  3.750  t   t    5   t   g  00100      3   +
4   b  20.17   5.625  u  g   w  v  1.710  t   f    0   f   s  00120      0   +
5   b  32.08   4.000  u  g   m  v  2.500  t   f    0   t   g  00360      0   +
6   b  33.17   1.040  u  g   r  h  6.500  t   f    0   t   g  00164  31285   +
7   a  22.92  11.585  u  g  cc  v  0.040  t   f    0   f   g  00080   1349   +
8   b  54.42   0.500  y  p   k  h  3.960  t   f    0   f   g  00180    314   +
9   b  42.50   4.915  y  p   w  v  3.165  t   f    0   t   g  00052   1442   +
10  b  22.08   0.830  u  g   c  h  2.165  f   f    0   t   g  00128      0   +
11  b  29.92   1.835  u  g   c  h  4.335  t   f    0   f   g  00260    200   +
12  a  38.25   6.000  u  g   k  v  1.000  t   f    0   t   g  00000      0   +
13  b  48.08   6.040  u  g   k  v  0.040  f   f    0   f   g  00000   2690   +
14  a  45.83  10.500  u  g   q  v  5.000  t   t    7   t   g  00000      0   +
15  b  36.67   4.415  y  p   k  v  0.250  t   t   10   t   g  00320      0   +
16  b  28.25   0.875  u  g   m  v  0.960  t   t    3   t   g  00396      0   +
17  a  23.25   5.875  u  g   q  v  3.170  t   t   10   f   g  00120    245   +
18  b  21.83   0.250  u  g   d  h  0.665  t   f    0   t   g  00000      0   +
19  a  19.17   8.585  u  g  cc  h  0.750  t   t    7   f   g  00096      0   +

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [0]:
# TODO
import scipy.stats as stats;

In [151]:
dfcheck1 = newdf[["A14", "A16"]]
dfcheck1plus = dfcheck1[dfcheck1["A16"] == "+"]
dfcheck1minus = dfcheck1[dfcheck1["A16"] == "-"]
print(dfcheck1plus.head(), "\n", dfcheck1minus.head())
print("plus", stats.ttest_1samp(np.array(dfcheck1plus["A14"], dtype=float), 0.5))
print("minus", stats.ttest_1samp(np.array(dfcheck1minus["A14"], dtype=float), 0.5))

     A14 A16
0  00202   +
1  00043   +
2  00280   +
3  00100   +
4  00120   + 
       A14 A16
70  00420   -
72  00980   -
73  00400   -
74  00160   -
75  00160   -
plus Ttest_1sampResult(statistic=17.371683792315853, pvalue=4.785690265525649e-47)
minus Ttest_1sampResult(statistic=21.184185239066558, pvalue=5.015664766131513e-65)


In [123]:
dfcheck2 = newdf[["A11", "A16"]]
dfcheck2plus = dfcheck2[dfcheck2["A16"] == "+"]
dfcheck2minus = dfcheck2[dfcheck2["A16"] == "-"]
print(dfcheck2plus.head(), "\n", dfcheck2minus.head())
print("plus", stats.ttest_1samp(list(dfcheck2plus["A11"]), 0.5))
print("minus", stats.ttest_1samp(list(dfcheck2minus["A11"]), 0.5))

   A11 A16
0    1   +
1    6   +
2    0   +
3    5   +
4    0   + 
     A11 A16
70    0   -
72    0   -
73    0   -
74    0   -
75    1   -
plus Ttest_1sampResult(statistic=11.337446996114545, pvalue=5.68150788132625e-25)
minus Ttest_1sampResult(statistic=1.608009129132978, pvalue=0.10871964179469769)


In [152]:
dfcheck3 = newdf[["A2", "A16"]]
dfcheck3plus = dfcheck3[dfcheck3["A16"] == "+"]
dfcheck3minus = dfcheck3[dfcheck3["A16"] == "-"]
print(dfcheck3plus.head(), "\n", dfcheck3minus.head())
print("plus", stats.ttest_1samp(np.array(dfcheck3plus["A2"], dtype=float), 0.5))
print("minus", stats.ttest_1samp(np.array(dfcheck3minus["A2"], dtype=float), 0.5))

      A2 A16
0  30.83   +
1  58.67   +
2  24.50   +
3  27.83   +
4  20.17   + 
        A2 A16
70  32.33   -
72  38.58   -
73  44.25   -
74  44.83   -
75  20.67   -
plus Ttest_1sampResult(statistic=45.21089737966061, pvalue=1.1529829659145896e-134)
minus Ttest_1sampResult(statistic=51.227426764703246, pvalue=2.4803241030093822e-166)


In [153]:
dfcheck4 = newdf[["A8", "A16"]]
dfcheck4plus = dfcheck4[dfcheck4["A16"] == "+"]
dfcheck4minus = dfcheck4[dfcheck4["A16"] == "-"]
print(dfcheck4plus.head(), "\n", dfcheck4minus.head())
print("plus", stats.ttest_1samp(list(dfcheck4plus["A8"]), 0.5))
print("minus", stats.ttest_1samp(list(dfcheck4minus["A8"]), 0.5))

     A8 A16
0  1.25   +
1  3.04   +
2  1.50   +
3  3.75   +
4  1.71   + 
         A8 A16
70   1.585   -
72  13.500   -
73  10.750   -
74   1.625   -
75   0.375   -
plus Ttest_1sampResult(statistic=12.282728352699618, pvalue=2.7564034680860017e-28)
minus Ttest_1sampResult(statistic=6.738568319848268, pvalue=6.453912635614688e-11)


In [147]:
table_obersved_1 = pd.crosstab(newdf["A16"], newdf["A1"])
table_obersved_1

A1     a    b
A16          
+     95  201
-    108  249

In [154]:
chi2, p, dof, expected = stats.chi2_contingency(table_obersved_1)
print("chi2:", chi2, "\np:", p, "\ndof:", dof, "\nexpected:", expected)

chi2: 0.17764082160253514 
p: 0.6734085695133722 
dof: 1 
expected: [[ 92.01837672 203.98162328]
 [110.98162328 246.01837672]]


In [155]:
table_obersved_2 = pd.crosstab(newdf["A16"], newdf["A5"])
table_obersved_2

A5     g  gg    p
A16              
+    249   2   45
-    250   0  107

In [156]:
chi2, p, dof, expected = stats.chi2_contingency(table_obersved_2)
print("chi2:", chi2, "\np:", p, "\ndof:", dof, "\nexpected:", expected)

chi2: 21.78325079317282 
p: 1.8613463470618034e-05 
dof: 2 
expected: [[226.19295559   0.90658499  68.90045942]
 [272.80704441   1.09341501  83.09954058]]


In [157]:
table_obersved_3 = pd.crosstab(newdf["A12"], newdf["A5"])
table_obersved_3

A5     g  gg   p
A12             
f    269   0  82
t    230   2  70

In [158]:
chi2, p, dof, expected = stats.chi2_contingency(table_obersved_3)
print("chi2:", chi2, "\np:", p, "\ndof:", dof, "\nexpected:", expected)

chi2: 2.3317179636969794 
p: 0.31165484117092523 
dof: 2 
expected: [[268.22205207   1.07503828  81.70290965]
 [230.77794793   0.92496172  70.29709035]]


## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

*Your words here!*

When running the ttest I used a continues variable where the range of values were very different for each class, and one where the range difference was smaller in comparison. 

I first ran the test for the smaller range difference and found the difference in statistics to be around 4 units and the difference in p value to be around 1 unit.
When I ran the test for the variable with a bigger range, I noticed that the difference in the in the statistics was around 10 units and the difference in the p value was around 5 units.

The second ttest shows that the class can greatly affect the vlaues of A11. But the first test shows the class has a smaller effect on the values of A14.

The chi-squared test with A1 produced a small chi value of 0.18 and a p value of 0.67. The small value indicated that the class label has a very small impact on the A1 values

The chi-squared test with A5 produced a larger chi value of 21.78 and a bigger p value of 1.87. These larger values indicate that the values of A5 are much different depending on the class label. This means the certain values will appear more frequently with one class and less frequently with the other class. This rejects the null hypothesis as the data values of A5 are dependent on the class label in A16.

The most challenging part was finding the right data to test with to find varying results from running the tests. I had to test different values to find results that had a big enough difference to work with. Then I had to figure out what the difference meant, as in, whether or not they were dependent on the class or not